File: experimental/teams/js/mentionsBot/src/mentionsBot.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import {
    Mention,
    MessageFactory,
    TeamsActivityHandler,
} from 'botbuilder';

export class MentionsBot  extends TeamsActivityHandler {
    /*
     * You can @mention the bot from any scope and it will reply with the mention.
     */    
    constructor() {
        super();

        // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
        this.onMessage(async (context, next) => {
            var mention = { mentioned: context.activity.from, text:`<at>${context.activity.from.name}</at>` };

            // Against Teams having a Mention in the Entities but not including that
            // mention Text in the Activity Text will result in a BadRequest.
            var replyActivity = MessageFactory.text(`Hello ${mention.text}.`);
            replyActivity.entities = [ <Mention> mention ];

            await context.sendActivity(replyActivity);

            // By calling next() you ensure that the next BotHandler is run.
            await next();
        });
    }
}

File: experimental/teams/csharp-dotnetcore/Mentions/Bots/MentionsBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples.Bots
{
    public class MentionsBot : ActivityHandler
    {
        /*
         * You can @mention the bot from any scope and it will reply with the mention.
         */
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var mention = new Mention
            {
                Mentioned = turnContext.Activity.From,
                Text = $"<at>{turnContext.Activity.From.Name}</at>",
            };

            // Against Teams having a Mention in the Entities but not including that
            // mention Text in the Activity Text will result in a BadRequest.
            var replyActivity = MessageFactory.Text($"Hello {mention.Text}.");
            replyActivity.Entities = new List<Entity> { mention };

            await turnContext.SendActivityAsync(replyActivity, cancellationToken);
        }
    }
}
